### ISSUE TITLE
AuthorizeUri - Insufficient Permissions

### ISSUE THREAD

**Issue by victorlee1** [ringcentral/ringcentral-csharp#49]
I am using the official C# SDK. And in order to get the URL I call: Platform.AuthorizeUri({redirect_uri});

The above call returns me the URL:

https://platform.devtest.ringcentral.com/restapi/oauth/authorize?response_type=code&state=&client_id=BrxnOxGBQgaeRYjbAYX25w&redirect_uri=http://localhost:8015/web/caller/Dialer_RingCentral.asp?step=2

I verified that:

- The **client_id** is indeed my RingCentral App key
- The **redirect_uri** is correctly configured in my RingCentral App

I then open the above URL and get the RingCentral's credentials window. Finally, after clicking the "authorize" button I get the following message:

![image](https://cloud.githubusercontent.com/assets/20427193/16782978/1cc577f8-4850-11e6-8aed-d812707df849.png)

I can't find what the problem is. I have already contacted RingCentral's dev support and they ran the PHP sample on their end (using my Apps credentials) and it seems to work just fine.

I am lost here.

**Comment by tylerlong** [comment]
It doesn't seem like an SDK issue. I would like to have a try. Do you mind sharing your my Apps credentials with me?  tyler.liu@ringcentral.com  It is even better if you also send me the sample code you used to do the testing.

**Comment by tylerlong** [comment]
@victorlee1, by the way, do you know there is a demo for OAuth: https://github.com/grokify/ringcentral-demos-oauth/tree/master/csharp-nancy ?

**Comment by tylerlong** [comment]
I tried you credentials with https://tylerlong.github.io/rc-api-explorer/ and it worked so I think there is no problem with the credentials.
Let me do more tests.

**Comment by tylerlong** [comment]
I tried your credentials with https://github.com/grokify/ringcentral-demos-oauth/tree/master/csharp-nancy and it worked. 

I am going to investigate the following:

1. The redirect url I used is http://localhost:8080/redirect.html, but yours is http://localhost:8015/web/caller/Dialer_RingCentral.asp?step=2  There are special characters in your url which might be the root cause.
2. Do you mind sending me your appKey and appSecret?

**Comment by victorlee1** [comment]
Hi Tyler,

thanks for all the effort. I just downloaded the csharp nancy version and I
am going to compare with my implementation. I might have been missing
something...

In anyways, I will update my ticket in github once I find what the problem
is.

Thank you!

On Tue, Jul 12, 2016 at 11:33 PM, Tyler Long <notifications@github.com>
wrote:

> I tried your credentials with
> https://github.com/grokify/ringcentral-demos-oauth/tree/master/csharp-nancy
> and it worked.
>
> I am going to investigate the following:
>
>    1. The redirect url I used is http://localhost:8080/redirect.html, but
>    yours is http://localhost:8015/web/caller/Dialer_RingCentral.asp?step=2
>    There are special characters in your url which might be the root cause.
>    2. Do you mind sending me your appKey and appSecret?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ringcentral/ringcentral-csharp/issues/49#issuecomment-232247928>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ATexuSQMqj50YIvn5iSzbkyvoBa1FDT_ks5qVFyhgaJpZM4JKz2a>
> .
>



-- 
*Victor Simon Lee*
Software Developer

**Comment by victorlee1** [comment]
Hi Tyler, thanks for all the effort. I just downloaded the csharp nancy
version and I am going to compare with my implementation. I might have been
missing something... In anyways, I will write a comment in my ticket once I
find what the problem is. Thank you!

On Tue, Jul 12, 2016 at 11:33 PM, Tyler Long <notifications@github.com>
wrote:

> I tried your credentials with
> https://github.com/grokify/ringcentral-demos-oauth/tree/master/csharp-nancy
> and it worked.
>
> I am going to investigate the following:
>
>    1. The redirect url I used is http://localhost:8080/redirect.html, but
>    yours is http://localhost:8015/web/caller/Dialer_RingCentral.asp?step=2
>    There are special characters in your url which might be the root cause.
>    2. Do you mind sending me your appKey and appSecret?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ringcentral/ringcentral-csharp/issues/49#issuecomment-232247928>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ATexuSQMqj50YIvn5iSzbkyvoBa1FDT_ks5qVFyhgaJpZM4JKz2a>
> .
>



-- 
*Victor Simon Lee*
Software Developer

**Comment by victorlee1** [comment]
Okay, so I figured out what the problem was.

I happens that the redirect_uri cannot contain any parameters. Seems it is an OAuth requirement?

So, after removing the queystring parameters from my return_uri, it all worked as expected.

Thanks for the help!

**Comment by tylerlong** [comment]
Actually I tried redirect_uri with parameters and I managed to make it work. Something to pay attention to:

1. You might need to `Uri.EscapeDataString(redirect_uri)`
2. In Step 2 of this guide: http://ringcentral-csharp.readthedocs.io/en/latest/usage/authorization/Authorization/ . `redirectUri`'s parameters should be removed before call `sdk.Platform.Authenticate(authCode, redirectUri)`

I don't know it's the requirement of OAuth specification or just the requirement of RingCentral OAuth or even a mistake we have on the API server side. More investigation needed here. I hope the information above helps :)
